### ISSUE TITLE
Expose XHR

### ISSUE THREAD

**Issue by kmiyashiro** [ember-cli/ember-ajax#67]
I was trying to follow the example pattern of aborting XHRs when retrying a task in ember-concurrency (http://ember-concurrency.com/#/docs/examples/autocomplete), but realized that there's no way to abort a request made by ember-ajax. Is exposing the underlying raw XHR per request a possibility?

**Comment by webark** [comment]
@kmiyashiro you can make a "raw" request rather then just a "request".

**Comment by alexlafroscia** [comment]
@webark has it right; using `raw` will give you the XHR back.

We're going to explore better ways to make `ember-ajax` play nicely with `ember-concurrency`, but we would need to implement cancellable promises and are just not quite ready for that right now.

**Comment by XaserAcheron** [comment]
Late response, but @machty is right -- the suggested workaround doesn't exist without reimplementing (read: copy+pasting) the `raw` method entirely, as there's no way to get to the underlying XHR since it's wrapped in a promise.

I'm not really sure what to suggest as an alternative. `Ember.RSVP.defer` provides a way to force-reject/resolve a promise, but its use isn't recommended and "abort" is a separate concept from "resolve" and "reject" anyway.

re:cancellable promises: Think this is worth feature-suggesting on RSVP's side? That'd naturally be the "sanest" API for doing something like this.

**Comment by tabeth** [comment]
I can also corroborate what @XaserAcheron and @machty say. For now, I've simply begun to use jQuery directly, which isn't ideal (given the purpose of `ember-ajax`, but it provides a nicer interface with `ember-concurrency`). 

I'd say that this issue should be reopened, or another issue should be created. AFAIK it's not possible to use `ember-ajax` with `ember-concurrency`, without significant modifications.

**Comment by rwjblue** [comment]
Agreed. We should probably return a custom thenable that allows access to the raw XHR before being resolved. As long as we ultimately resolve the same way we do today, it is not a breaking change...

**Comment by XaserAcheron** [comment]
Diving into the code a bit, a simple-ish way to do this is to extend the Promise returned by [`_makeRequest`](https://github.com/ember-cli/ember-ajax/blob/master/addon/mixins/ajax-request.js#L233) to add a `jqXHR` property. Currently doing some code spelunking to determine the best way to go about such a thing.

Looking at the code, though, I'm a bit confused by the implementations of `raw` and `request`: The former returns the `_makeRequest` promise directly, while [the latter](https://github.com/ember-cli/ember-ajax/blob/master/addon/mixins/ajax-request.js#L197) just seems to wrap the result of `_makeRequest` in a second promise that effectively does nothing but change the Promise label -- i.e. `raw` and `request` do exactly the same thing. I suspect the intent was for `request` to return a Promise and for `raw` to return the `jqXHR` object directly, but something was lost in translation at some point.

I bring this up to ask a related question: If we go the custom-thenable route, is there any reason we _shouldn't_ allow access to the `jqXHR` from `request` calls? If that's palatable, I'd propose attaching the property to `request`'s promise and deprecate `raw` entirely, as it would be a nuisance to have to skip all the handy `get`, `post`, etc. helpers to call `raw` just to be able to do this.

[Wow, lots of words.]

**Comment by taras** [comment]
> We should probably return a custom thenable that allows access to the raw XHR before being resolved.

@rwjblue We didn't think of this solution. Thank you for suggesting it.

> If we go the custom-thenable route, is there any reason we shouldn't allow access to the jqXHR from request calls?

This seems to be a legit path to me.

> If that's palatable, I'd propose attaching the property to request's promise

Sounds good to me.

> deprecate raw entirely

We can do this in stages. 
1. `jqXHR` to the promise
2. Create a proposal issue to deprecate `raw`
3. Deprecate `raw` depending on feedback that we get

**Comment by taras** [comment]
FYI: we have Ember Ajax Working Group meeting on Wednesday at 12:00 PM EST. Let me know if anyone would like an invite to participate.

**Comment by XaserAcheron** [comment]
Is it voice or text? I'm interested in sitting in (if outside folk are welcome :P ), but it's during the workday for me so I'll have to see if I can schedule some time first.

[I'm psyched to see folks get involved in this, BTW. Thanks for the input!]

**Comment by taras** [comment]
> Is it voice or text? 
It'll be voice.

> I'm interested in sitting in (if outside folk are welcome :P ), but it's during the workday for me so I'll have to see if I can schedule some time first.

Everyone is welcome to join us.

> [I'm psyched to see folks get involved in this, BTW. Thanks for the input!]

Thank you for restarting this conversation.

**Comment by XaserAcheron** [comment]
I may have some time free after all, if it's not too late for the invite.

**Comment by taras** [comment]
@XaserAcheron can you please send me an email taras@embersherpa.com

**Comment by robclancy** [comment]
Has there been any progress on this?

**Comment by alexlafroscia** [comment]
@robclancy not yet. As far as I know, we need to decide on whether or not we're keeping `ember-ajax` a wrapper for `$.ajax`, or if we want it to ever become a `fetch` wrapper instead (my vote: stick with jQuery). Once we know that, we'll work out a method for exposing the XHR, but if we move to `fetch` then there wouldn't be an XHR to expose.

**Comment by robclancy** [comment]
It's been open since March and there is still no timeframe on knowing what is happening in the future? What harm can supporting this have?

This is the way I see it. If you are sticking with `$.ajax` and do this change nothing happens, it just remains. If you need to change to use `fetch` you will still be tagging a "last release" that supports `$.ajax` anyway.

Also another option is simply making it easier to override. Right now to override and expose I need to copy paste the entire `_makeRequest` method into my apps `ajax` service, I guess might make sense for me to just fork for now.

Is there somewhere to track all future stuff? Is it just a part of an ember RFC or something?

**Comment by alexlafroscia** [comment]
> It's been open since March and there is still no timeframe on knowing what is happening in the future? What harm can supporting this have?

I, personally, don't have any interest in adding a feature that we'll later be removing. Up until very recently, I pretty much did all of the work on bugs/features here myself, along with @taras and a few others that helped with some maintenance task. Since March I graduated college, moved across the country, and started a new job. I'm sorry that this Github issue is not of particular import. Now, we have a group that meets bi-weekly to discuss issues, plan future development, and make decisions on how we want this add-on to develop. If you're interested in joining us, I'd be happy to send you the details.

> Also another option is simply making it easier to override. Right now to override and expose I need to copy paste the entire _makeRequest method into my apps ajax service, I guess might make sense for me to just fork for now.

If you would like to make a pull request with your proposed solution for making this easier to override, I would love to help get it into the library!

> Is there somewhere to track all future stuff? 

We actually have what needs to happen for 3.0.0 planned out pretty nicely in the [3.0.0 project](https://github.com/ember-cli/ember-ajax/projects/1). As you can see, determining what to do about being an `$.ajax` wrapper or moving to `fetch` for 4.0.0 is on there. If you want to help us talk through the details, as I said before, I'm happy to send you the details on our bi-weekly meeting!

> Is it just a part of an ember RFC or something?

Not yet, but we're trying to determine the right place to make RFCs on high-profile addons. Really, the only one that does them right now is `ember-data`, and the community as a whole hasn't decided where to organize them. However, this is something that we've been talking about, so you can expect to see a `fetch` vs `$.ajax` RFC somewhere very soon.
